File: image_viewer.cpp
// image_viewer.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <string>
#include <filesystem>
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#include <algorithm>

using namespace std;
using namespace cv;

namespace fs = std::filesystem;

string RESET = "\033[0m";
string BOLD = "\033[1m";

string rgbToAnsi(int r, int g, int b) {
    return "\033[38;2;" + to_string(r) + ";" + to_string(g) + ";" + to_string(b) + "m";
}

vector<string> getImageFiles(const string& path) {
    vector<string> files;
    if (fs::is_regular_file(path)) {
        files.push_back(path);
    } else if (fs::is_directory(path)) {
        vector<string> exts = {".jpg", ".jpeg", ".png", ".gif", ".bmp", ".tiff", ".webp"};
        for (const auto& entry : fs::directory_iterator(path)) {
            string ext = entry.path().extension().string();
            transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
            if (find(exts.begin(), exts.end(), ext) != exts.end()) {
                files.push_back(entry.path().string());
            }
        }
        sort(files.begin(), files.end());
    }
    return files;
}

char getch() {
    struct termios oldt, newt;
    char ch;
    tcgetattr(STDIN_FILENO, &oldt);
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
    ch = getchar();
    tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
    return ch;
}

int getTerminalWidth() {
    struct winsize w;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
    return w.ws_col;
}

int getTerminalHeight() {
    struct winsize w;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
    return w.ws_row;
}

void displayImage(const Mat& img, int zoom, int rotate, bool flipH, bool flipV, bool asciiMode) {
    Mat display = img.clone();
    if (rotate == 90) rotate(display, display, ROTATE_90_CLOCKWISE);
    else if (rotate == 180) rotate(display, display, ROTATE_180);
    else if (rotate == 270) rotate(display, display, ROTATE_90_COUNTERCLOCKWISE);
    if (flipH) flip(display, display, 1);
    if (flipV) flip(display, display, 0);

    int w = display.cols, h = display.rows;
    int termW = getTerminalWidth() - 2;
    int termH = getTerminalHeight() - 6;
    double scale = zoom;
    if (scale == 1.0) {
        double scaleX = (double)termW / w;
        double scaleY = (double)termH / h;
        scale = min(scaleX, scaleY);
        if (scale > 1.0) scale = 1.0;
    }
    int newW = max(1, (int)(w * scale));
    int newH = max(1, (int)(h * scale));
    Mat resized;
    resize(display, resized, Size(newW, newH), 0, 0, INTER_LANCZOS4);

    cout << "\033[H\033[2J";
    cout << BOLD << "📷 Image Viewer | " << newW << "×" << newH << BOLD << RESET << endl;
    cout << "Zoom: " << zoom << "x | Rotate: " << rotate << "° | Flip: " << flipH << "/" << flipV << endl;
    cout << "Controls: ← → (nav)  +/- (zoom)  r (rotate)  h/v (flip)  f (mode)  q (quit)" << endl;

    if (asciiMode) {
        Mat gray;
        cvtColor(resized, gray, COLOR_BGR2GRAY);
        string chars = " .:-=+*#%@";
        for (int r = 0; r < gray.rows; ++r) {
            for (int c = 0; c < gray.cols; ++c) {
                int val = gray.at<uchar>(r, c);
                int idx = val * (chars.size()-1) / 255;
                cout << chars[idx];
            }
            cout << endl;
        }
    } else {
        for (int r = 0; r < resized.rows; ++r) {
            for (int c = 0; c < resized.cols; ++c) {
                Vec3b pixel = resized.at<Vec3b>(r, c);
                cout << rgbToAnsi(pixel[2], pixel[1], pixel[0]) << "██";
            }
            cout << RESET << endl;
        }
    }
}

int main(int argc, char* argv[]) {
    string path = ".";
    double zoom = 1.0;
    int rotate = 0;
    bool asciiMode = false;
    string convertFormat, outputFile;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "-z" && i+1 < argc) zoom = stod(argv[++i]);
        else if (arg == "-r" && i+1 < argc) rotate = stoi(argv[++i]);
        else if (arg == "-a") asciiMode = true;
        else if (arg == "-c" && i+1 < argc) convertFormat = argv[++i];
        else if (arg == "-o" && i+1 < argc) outputFile = argv[++i];
        else if (arg == "-h" || arg == "--help") {
            cout << "Usage: image_viewer <path> [-z zoom] [-r rotate] [-a] [-c format] [-o output]" << endl;
            return 0;
        } else if (path == ".") path = arg;
    }

    auto files = getImageFiles(path);
    if (files.empty()) {
        cerr << "Нет изображений." << endl;
        return 1;
    }

    int idx = 0;
    bool flipH = false, flipV = false;
    Mat img = imread(files[idx]);
    if (img.empty()) { cerr << "Не удалось загрузить " << files[idx] << endl; return 1; }

    if (!convertFormat.empty() && !outputFile.empty()) {
        imwrite(outputFile, img);
        cout << "✅ Изображение сохранено в " << outputFile << endl;
        return 0;
    }

    while (true) {
        displayImage(img, zoom, rotate, flipH, flipV, asciiMode);
        char key = getch();
        if (key == 'q') break;
        else if (key == '+') zoom = min(3.0, zoom + 0.2);
        else if (key == '-') zoom = max(0.2, zoom - 0.2);
        else if (key == 'r') { rotate = (rotate + 90) % 360; }
        else if (key == 'h') { flipH = !flipH; }
        else if (key == 'v') { flipV = !flipV; }
        else if (key == 'f') { asciiMode = !asciiMode; }
        else if (key == 27) { // ESC
            char seq[2];
            if (getchar() == '[') {
                char c = getchar();
                if (c == 'D' && idx > 0) { idx--; img = imread(files[idx]); }
                else if (c == 'C' && idx < (int)files.size()-1) { idx++; img = imread(files[idx]); }
            }
        }
    }
    return 0;
}

File: image_viewer.go
// image_viewer.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"image"
	"image/color"
	_ "image/gif"
	_ "image/jpeg"
	_ "image/png"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/disintegration/imaging"
	"golang.org/x/term"
)

const (
	reset = "\033[0m"
	bold  = "\033[1m"
)

func rgbToAnsi(r, g, b uint8) string {
	return fmt.Sprintf("\033[38;2;%d;%d;%dm", r, g, b)
}

func getImageFiles(path string) []string {
	var files []string
	info, err := os.Stat(path)
	if err != nil {
		return files
	}
	if !info.IsDir() {
		return []string{path}
	}
	exts := map[string]bool{".jpg": true, ".jpeg": true, ".png": true, ".gif": true, ".bmp": true, ".tiff": true, ".webp": true}
	entries, _ := os.ReadDir(path)
	for _, e := range entries {
		if !e.IsDir() {
			ext := strings.ToLower(filepath.Ext(e.Name()))
			if exts[ext] {
				files = append(files, filepath.Join(path, e.Name()))
			}
		}
	}
	sort.Strings(files)
	return files
}

func displayImage(img image.Image, zoom float64, rotate int, flipH, flipV bool, asciiMode bool, termW, termH int) {
	// Применяем трансформации
	var processed image.Image = img
	if rotate == 90 {
		processed = imaging.Rotate90(processed)
	} else if rotate == 180 {
		processed = imaging.Rotate180(processed)
	} else if rotate == 270 {
		processed = imaging.Rotate270(processed)
	}
	if flipH {
		processed = imaging.FlipH(processed)
	}
	if flipV {
		processed = imaging.FlipV(processed)
	}
	// Масштабирование
	bounds := processed.Bounds()
	w, h := bounds.Dx(), bounds.Dy()
	maxW := termW - 2
	maxH := termH - 6
	scale := zoom
	if scale == 1.0 {
		scaleX := float64(maxW) / float64(w)
		scaleY := float64(maxH) / float64(h)
		scale = min(scaleX, scaleY)
		if scale > 1.0 {
			scale = 1.0
		}
	}
	newW := int(float64(w) * scale)
	newH := int(float64(h) * scale)
	if newW < 1 {
		newW = 1
	}
	if newH < 1 {
		newH = 1
	}
	resized := imaging.Resize(processed, newW, newH, imaging.Lanczos)

	fmt.Print("\033[H\033[2J")
	fmt.Printf("%s📷 Image Viewer | %d×%d%s\n", bold, newW, newH, reset)
	fmt.Printf("Zoom: %.1fx | Rotate: %d° | Flip: %t/%t\n", zoom, rotate, flipH, flipV)
	fmt.Println("Controls: ← → (nav)  +/- (zoom)  r (rotate)  h/v (flip)  f (mode)  q (quit)")

	if asciiMode {
		chars := " .:-=+*#%@"
		for y := 0; y < newH; y++ {
			for x := 0; x < newW; x++ {
				c := resized.At(x, y)
				r, g, b, _ := c.RGBA()
				gray := 0.299*float64(r>>8) + 0.587*float64(g>>8) + 0.114*float64(b>>8)
				idx := int(gray / 255.0 * float64(len(chars)-1))
				fmt.Print(string(chars[idx]))
			}
			fmt.Println()
		}
	} else {
		for y := 0; y < newH; y++ {
			for x := 0; x < newW; x++ {
				c := resized.At(x, y)
				r, g, b, _ := c.RGBA()
				fmt.Print(rgbToAnsi(uint8(r>>8), uint8(g>>8), uint8(b>>8)) + "██")
			}
			fmt.Println(reset)
		}
	}
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func getch() byte {
	oldState, _ := term.MakeRaw(int(os.Stdin.Fd()))
	defer term.Restore(int(os.Stdin.Fd()), oldState)
	var buf [1]byte
	os.Stdin.Read(buf[:])
	return buf[0]
}

func main() {
	var path string
	var zoom float64
	var rotate int
	var asciiMode bool
	var convertFormat string
	var outputFile string
	flag.StringVar(&path, "path", ".", "Файл или папка")
	flag.Float64Var(&zoom, "z", 1.0, "Масштаб")
	flag.IntVar(&rotate, "r", 0, "Поворот")
	flag.BoolVar(&asciiMode, "a", false, "ASCII-режим")
	flag.StringVar(&convertFormat, "c", "", "Конвертировать в формат")
	flag.StringVar(&outputFile, "o", "", "Выходной файл")
	flag.Parse()
	if flag.NArg() > 0 {
		path = flag.Arg(0)
	}

	files := getImageFiles(path)
	if len(files) == 0 {
		fmt.Println("Нет изображений.")
		return
	}

	idx := 0
	flipH, flipV := false, false
	img, err := imaging.Open(files[idx])
	if err != nil {
		fmt.Println("Ошибка загрузки:", err)
		return
	}

	if convertFormat != "" && outputFile != "" {
		err := imaging.Save(img, outputFile)
		if err == nil {
			fmt.Printf("✅ Изображение сохранено в %s\n", outputFile)
		}
		return
	}

	termW, termH, _ := term.GetSize(int(os.Stdin.Fd()))

	for {
		displayImage(img, zoom, rotate, flipH, flipV, asciiMode, termW, termH)
		key := getch()
		switch key {
		case 'q':
			return
		case '+':
			if zoom < 3.0 {
				zoom += 0.2
			}
		case '-':
			if zoom > 0.2 {
				zoom -= 0.2
			}
		case 'r':
			rotate = (rotate + 90) % 360
		case 'h':
			flipH = !flipH
		case 'v':
			flipV = !flipV
		case 'f':
			asciiMode = !asciiMode
		case 27:
			// стрелка
			buf := make([]byte, 2)
			os.Stdin.Read(buf)
			if buf[0] == '[' {
				if buf[1] == 'D' && idx > 0 {
					idx--
					img, _ = imaging.Open(files[idx])
				} else if buf[1] == 'C' && idx < len(files)-1 {
					idx++
					img, _ = imaging.Open(files[idx])
				}
			}
		}
	}
}
